### PR TITLE
Add support for re2c parser

### PR DESCRIPTION
**Pull request by alemuller**
The re2c is a lexer generator. It is a language that only exists inside comment blocks.

The language can be injected into C (and other languages that support C syntax), Go and Rust.

Examples:
* [re2c/examples/c/*.re](https://github.com/skvadrik/re2c/tree/master/examples/c)
* [ninja/src/*.in.cc](https://github.com/ninja-build/ninja/blob/master/src/)

The official repo uses `.re` extensions (not only for `.c`, but for other filetypes as well). Use `set ft=c` to highlight. 

I have some issues.
1. The re2c should inject the `filetype` it is being injected on. For example, when re2c is being injected in a c file, the `(code_block)` should be injected with `@c`. If the file is .go, it should inject `@go`. And so on.
2. I also have issues writing the indentation when there is a injected block. Try to indent try to indent [depfile_parser.in.cc](https://github.com/ninja-build/ninja/blob/master/src/depfile_parser.in.cc)) to see what happens.
3. I couldn't get folding to work. I'd like to know if is possible to fold sequence of the same blocks. Something like `(configuration)+ @fold`.

How I can fix those issues?

Because the first issue, I only wrote injections for c and cpp so you could see the issues I'm having with the examples above.

---

Some other notes.

The re2c parser expects the input not to include `/*` and `*/`. So, if using `set ft=re2c` start with the exclamation point. They were not included in the grammar saves some kb and improves the error recovery. 

re2c uses the same comment syntax from C. Comments inside `(code_block) are parsed as re2c comments (easier to write the rule that seeks the matching closing brace -- that works really well, btw).

The `(ignore_block)` has a greedy regex. The official lexer matches `[^\0]`. My version don't match the '*/' to make sure it is not going over the C file (even thought this is probably not an issue).

### CODE REVIEW

**Comment by alemuller** [comment]
I'll try something else.

Instead of injecting in C comments. I will change the re2c grammar to matching the surround of comments and inject c/go/rust there. So it won't need to change others languages queries.

The new AST would look something like:
```scm
(re2c
  (host_lang)
  (global_block ... (host_lang) ... (host_lang)) ; the re2c code
  (host_lang))
```

I will try to use the `first-line-regex` from the `package.json` to detect the languages on the injection query.

This would solve the issue of the injected language being the same as the injected language.

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -202,6 +202,16 @@ list.graphql = {
...
+list.re2c = {
+  install_info = {
+    url = "https://github.com/alemuller/tree-sitter-re2c",
+    files = { "src/parser.c" },
+    branch = "main",
+  },
+  maintainers = { "@alemuller" },
+}
+
+
```
Comment (theHamsta) [inline lua/nvim-treesitter/parsers.lua:214]: Can you keep the spacing between the parsers consistent?

**Comment by alemuller** [comment]
I getting an assertion violation when editing a file. Should I open an issue?

`01_basic.re`
```c
#include <stdio.h>

int lex(const char *YYCURSOR) {       // C/C++ code
/*!re2c
    re2c:define:YYCTYPE = char;       // configurations
    re2c:yyfill:enable  = 0;          //
                                      //
    alpha = [a-zA-Z];                 // named definitions
    ident = alpha (alpha | [0-9])*;   //
                                      //
    ident { return 0; }               // rules
    *     { return 1; }               //
*/
}                                     //
                                      //
int main() {                          // C/C++ code
    return lex("qwerty42");           //
}                                     //
```
Copying and pasting `/*!re2c` (`gg3jyyP`) causes nvim to crash.

**Comment by theHamsta** [comment]
@alemuller yes, please open an issue. The might be similar to the segfault on (couldn't find the Neovim issue)


```c
❯ cat foo.c
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
1: A.X_Obj.B.C.2.Enable: true
```
Did you do some GDB debugging for the origin of the SEGFAULT?

**Comment by alemuller** [comment]
The issue seems has been fixed on one of the last changed made to neovim. I didn't debug.

I'm not sure if injecting the host the language on the whole grammar is a good idea.

I'm thinking about reverting the grammar to only parse the content inside the delimiters (instead of parsing the file into host_lang/and re2c blocks), and use the `ftdetect/re2c.vim` to: identify the host language; change `&filetype`; and somehow load the correct injection query. I'm not sure where to store the injections, thought. Maybe `nvim-treesitter/queries/re2c/c/injections.c`.

I'd like to hear your thoughts,  @theHamsta.

**Pull request closed without merging**